src/AWS/Shared/Storage/DynamoDBStorage.cs

```diff
@@ -163,34 +165,191 @@ private async Task CreateTable(string tableName, List<KeySchemaElement> keys, Li
+                Logger.Error(ErrorCode.StorageProviderBase, $"Could not update table {tableDescription.TableName}", exc);
                 throw;
             }
         }
 
+        private async Task TableCreateSecondaryIndex(string tableName, List<AttributeDefinition> attributes, GlobalSecondaryIndex secondaryIndex)
+        {
+            await ddbClient.UpdateTableAsync(new UpdateTableRequest
+            {
+                TableName = tableName,
+                GlobalSecondaryIndexUpdates = new List<GlobalSecondaryIndexUpdate>
+                {
+                    new GlobalSecondaryIndexUpdate
+                    {
+                        Create = new CreateGlobalSecondaryIndexAction()
+                        {
+                            IndexName = secondaryIndex.IndexName,
+                            Projection = secondaryIndex.Projection,
+                            ProvisionedThroughput = provisionedThroughput,
+                            KeySchema = secondaryIndex.KeySchema
+                        }
+                    }
+                },
+                AttributeDefinitions = attributes
+            });
+
+            // Adding a GSI to a table is an eventually consistent operation and we might miss the table UPDATING status if we query the table status imediatelly after the table update call.
+            // Creating a GSI takes significantly longer than 1 second and therefore this delay does not add time to the total duration of this method.
+            await Task.Delay(1000);
+
+            // When adding a GSI, the table briefly changes its status to UPDATING. The GSI creation process usually takes longer.
+            // For this reason, we will wait for both the table and the index to become ACTIVE before marking the operation as complete.
+            await TableWaitOnStatusAsync(tableName, TableStatus.UPDATING, TableStatus.ACTIVE);
+            await TableIndexWaitOnStatusAsync(tableName, secondaryIndex.IndexName, IndexStatus.CREATING, IndexStatus.ACTIVE);
+        }
+
+        private async ValueTask TableUpdateTtlAsync(string tableName, string ttlAttributeName)
+        {
+            var describeTimeToLive = (await ddbClient.DescribeTimeToLiveAsync(tableName)).TimeToLiveDescription;
+
+            // We can only handle updates to the table TTL from DISABLED to ENABLED.
+            // This is because updating the TTL attribute requires (1) disabling the table TTL and (2) re-enabling it with the new TTL attribute.
+            // As per the below details page for this API: "It can take up to one hour for the change to fully process. Any additional UpdateTimeToLive calls for the same table during this one hour duration result in a ValidationException."
+            // https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_UpdateTimeToLive.html
+            if (describeTimeToLive.TimeToLiveStatus != TimeToLiveStatus.DISABLED)
```

Comment:
Going from enabled to disabled seems feasible, but it definitely goes beyond this PR's scope.

Reply:
Indeed, definetly possible. The logic however gets quite complicated. I think the infrastructure provisioning code should be converted to leverage IaC (CDK or Terraform CDK would work well here). This would significantly reduce complexity.